### ISSUE TITLE
Is it possible to save checkpoints without resetting/clearing the state of the model?

### ISSUE THREAD

**Issue by AlekzNet** [jcjohnson/torch-rnn#46]
From what I understand, without resetting the internal states "torch.save" saves everything, so the checkpoint file size can be gigabytes in size. However, char-rnn does it somehow (I'm currently trying to understand how it's done, but I'm not sure it's compatible with torch-rnn ).

Is it possible to implement a similar way of saving in torch-rnn?

**Comment by jcjohnson** [comment]
We get small checkpoints in torch-rnn by calling the `clearState` method, which is a pretty confusing name in the context of RNNs - `clearState` gets rid of intermediate activations and all gradients, so that checkpoints will be smaller. `clearState` does NOT get rid of the saved hidden state used for backpropagation through time - those get cleared by calling `resetStates` instead.

Therefore the problem with resetting the hidden states is not due to saving, but is instead a problem with running the model on the validation data; we reset the hidden state before running on validation data since it doesn't make sense to use the hidden state from the training data to initialize the hidden state on the validation data.

The right solution would be to do something less lazy with the cached hidden states than just resetting them, so that you could flush the hidden states before running on validation data but then restore the hidden states from the training data after running on validation data. A simple way to do this would be to implement a stack of hidden states, so that you would have push/pop methods instead of a single reset method.

**Comment by AlekzNet** [comment]
Aha, thanks for the explanation! 

What I tried is to remove the validation code, including model:resetStates(). However, model:clearState() does not want to work without model:resetStates() and throws errors.

**Comment by AlekzNet** [comment]
In other words, model:clearState() clears something what breaks the next following optimization step. But It works fine after model:resetStates()

**Comment by dgcrouse** [comment]
Stale issue, closing
